### ISSUE TITLE
Fine-tune inference always produces same class prediction

### ISSUE THREAD

**Issue by shahshawaiz** [google/automl#930]
I am trying to train an Efficientdet-d1. Currently, when I try to do inferences I always get same prediction. I have tried "visualize tf-record" utility and their seems to be no problem with the dataset or annotations. I have also made sure that the training examples are properly shuffled during creation of tfRecord.

Following are some spefics of my project;

Dataset:
> Training Example: 100k+
> Validation Examples: 900+ (hand labeled examples)
> 18 classes, with ~5000+ training examples for each class

Training Session:
> Batch Size: 16
> Number of examples per epcoh: 11434  
> Trained Epochs:  ~117 (~84200 steps)

Results:
> total loss: 0.679
> box_loss: 0.0011
> cls_loss: 0.509
> det_loss: is 0.567

Unfortunately I couldnt set up tensorboard before starting training session hence I currently dont have the training history.

Training comand
```
python main.py --mode=train --train_file_pattern=dataset/x/tfrecord/train/train*.tfrecord --val_file_pattern=dataset/x/tfrecord/val/val-easy*.tfrecord --model_name=efficientdet-d1 --model_dir=efficientdet-d1_saved --ckpt=efficientdet-d1 --train_batch_size=16 --eval_batch_size=16 --eval_samples=1024 --num_examples_per_epoch=11434  --num_epochs=2000  --hparams=cfg/x_config.yaml
```

Config file;
```
num_classes: 19
var_freeze_expr: '(efficientnet|fpn_cells|resample_p6)'
label_map: {1: alpha, 2: bravo, ... 18: charlie}
regenerate_source_id: True
moving_average_decay: 0
use_keras_model: False
```

Clearly, the problem is in clear sight but i am missing it. I would like to know if this problem makes sense to anyone, or if someone had similar problem and was able to fix it

Thanks a bunch :)

**Comment by kartik4949** [comment]
Hi, @shahshawaiz 
1. 119 epochs seems pretty low, try with 300 epochs complete.
2. you can try training without var freeze.
3. if above doesnt work please try with smaller subset of data and try to overfitt the model, see if model learns something in training data :) (this seems to be highly unlikely.)
thanks

**Comment by ameyparanjape** [comment]
Facing a similar issue. @shahshawaiz how long did it take you to train ~117 epochs? Hardware specs? (I'm curious!), thanks!
For me, with batch size = 8, it is taking ~120 seconds per 100 steps (super slow!)

**Comment by shahshawaiz** [comment]
I tried some stuff since I posted this issue. I have a working model now, which seems to work quite well :)

- Previously, I had two RTX 2070s (I couldn't use both GPUs for training due to lack of support of NCCL drivers on windows). I think it took me some 18 hrs to train for 100 epochs. BTW my dataset had about 135k training examples. This produced no detections at all. 
- Then I had a system upgrade, now is use RTX 3090. This week, I did an experiment with 15k training examples, batch size 64, and num_examples_per_epoch set to 10921. It took me around 1hr 45 mins to train this model for 20 epochs. I was doing some hyperparamter optimization so I had to repeat this experiment 18 more times, and the trainnign time was almost the same.

![image](https://user-images.githubusercontent.com/4154691/110136046-82bc1a80-7dcf-11eb-8fb5-3a879936018b.png)

- However, the key to getting detections was a correction in dataset (and partly hyperparatmer optimization). Previsouly, I was using too many negative examples (training images with no annotations, to make my model prone to false positives). After removing all negatives, I noticed a clear difference. Now, I can correctly predict 149/164 (90.8%) validation examples. So I would suggest looking again at the dataset.

**Comment by ameyparanjape** [comment]
Thanks a ton @shahshawaiz for the elaborate description. My dataset is about ~200k images with single or multiple bbox. What was your observation on freezing layers? Did you freeze at all? Again thank you for the insights!

**Comment by shahshawaiz** [comment]
I did HP optimization on `freeze_exp` too. For me, retraining `fpn` and `resample_6` layer made clear difference. Retraining whole was very slow, and produced almost same results as freezing `efficientnet`.

![image](https://user-images.githubusercontent.com/4154691/110137927-aed89b00-7dd1-11eb-894f-d580a573ed28.png)
